[Model] IBM/NASA Prithvi Geospatial model

christian-pinto opened the pull request:
This pull request proposes a first integration of Prithvi, a non-language model to vLLM. This is related to the discussion we had in a previous RFC https://github.com/vllm-project/vllm/issues/11065.

Prithvi is a non-autoregressive model ingesting torch tensors in input and generating on (or a list of) torch tensors. As suggested in the discussion linked to the RFC, this can be achieved in vLLM by integrating the model as a Pooling model that allows extracting the hidden_states right away. This is exactly what I did.

This PR is only addressing offline inference at the moment. 


The below code snippet shows how interacting with the model is done at the moment


```python

model = LLM(model= "./model"), skip_tokenizer_init=True, dtype="float32")

input_data = torch.tensor(...)
location_coords = torch.tensor(...)
mm_data = {
    "pixel_values": input_data,
    "location_coords": location_coords
}

prompt = {
    "prompt_token_ids": [1],
    "multi_modal_data": mm_data
}

outputs = model.encode(prompt)

```

Even though I am disabling the tokenizer, vLLM still requires the "prompt" to provide input tokens. Here I am just faking the input tokens as they have no meaning for the model. All the relevant input data is passed using the multimodal interface.

The actual model architecture is instantiated using an open source framework called TerraTorch. Users willing to use this model must include `terratorch` in their environment.


This model was successfully tested on ARM64 CPU (Mac), x86 CPU and NVidia T4 GPUs.

A full demonstrator of the model at work is available [here](https://github.com/christian-pinto/prithvi-vllm-demo).

christian-pinto committed:
Added missing inter_data_list in cpu_model_runner.py

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
_process_multimodal: avoid decoding prompt when no tokenizer is initialized

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
PoolingRequestOutput.from_seq_group: allow a list of tensors to be returned

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
PoolingRequestOutput.from_seq_group: allow a list of tensors to be returned by the model

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Add Prithvi Geospatial model

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

Isotr0py commented on vllm/worker/cpu_model_runner.py:
```diff
@@ -167,6 +167,7 @@ def prepare(self,
         self.input_data = ModelInputForCPUBuilder.ModelInputData(
             self.runner.model_config.uses_mrope)
         self.att_metadata_builder.prepare()
+        self.inter_data_list = []
```
```suggestion
```
I think this is unused in CPU model runner.

christian-pinto replied:
True,

however in 

https://github.com/vllm-project/vllm/blob/c786e757fae4519256e4ef88a7d4f56c3339d14d/vllm/attention/backends/placeholder_attn.py#L308-L321

`inter_data_list` is referenced regardless of the type of builder used. I can make the change in there, to check whether the specific builder has that field. 

This was otherwise breaking while testing on CPU.

Isotr0py replied:
> I can make the change in there, to check whether the specific builder has that field.

We should modify in `vllm/vllm/attention/backends/placeholder_attn.py` there instead of `cpu_runner` since it's unused and a bit hacky.

christian-pinto replied:
exactly what I meant. Let me do that.

DarkLight1337 commented on vllm/inputs/preprocess.py:
```diff
@@ -254,14 +254,17 @@ def _process_multimodal(
         """
-        tokenizer_group = self.get_tokenizer_group()
-        tokenizer = tokenizer_group.get_lora_tokenizer(lora_request)
+        if not self.tokenizer:
+            tokenizer = None
+        else:
+            tokenizer_group = self.get_tokenizer_group()
+            tokenizer = tokenizer_group.get_lora_tokenizer(lora_request)
+            if isinstance(prompt, list):
+                prompt = tokenizer.decode(prompt)
```
We can remove this part now, I forgot to remove it when updating the async method.

christian-pinto replied:
Thanks, I will remove the decode call altogether.

DarkLight1337 replied:
Actually let me do it in another PR

DarkLight1337 commented:
Didn't know that linking to PRs also causes it to be closed 😅

DarkLight1337 commented:
To check the correctness of this model, can you update the following files?
- `tests/models/registry.py` for the CI to pass
- `tests/models/multimodal/processing/test_common.py` to test the HF processor
- A new file under a new directory `tests/models/vision_generation` to check the correctness against HF

christian-pinto committed:
Check for attribute inter_data_list in placeholder_attn

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
> To check the correctness of this model, can you update the following files?
> 
> * `tests/models/registry.py` for the CI to pass

At the current moment there is a model available on HF (https://huggingface.co/ibm-nasa-geospatial/Prithvi-EO-2.0-300M-TL-Sen1Floods11), however it requires some minor changes including a proper HF config.json since so far the model was loaded solely via terratorch (uses torch lightning) and has a different config file format (I have an example conversion to HF config [here](https://github.com/christian-pinto/prithvi-vllm-demo/blob/main/model/config.json)).
Would it be ok to temporarily mark the model with `is_available_online=False`  until we sort the config file on HF?

> * `tests/models/multimodal/processing/test_common.py` to test the HF processor

If you see from the model code, we do not directly take an image in input, while we currently pre-process the image and feed it to the model in "patches" that require no processing to be applied. So, generally one image is processed through a sequence of inferences depending on the size of the image itself.

Long story short, we do not have a HF processor. Hence, nothing to be tested here at the moment.
 
> * A new file under a new directory `tests/models/vision_generation` to check the correctness of the model against HF

Let me have a better look at this one.

DarkLight1337 commented:
> Would it be ok to temporarily mark the model with `is_available_online=False` until we sort the config file on HF?

Sure, that's fine.

> If you see from the model code, we do not directly take an image in input, while we currently pre-process the image and feed it to the model in "patches" that require no processing to be applied. So, generally one image is processed through a sequence of inferences depending on the size of the image itself.

Hmm, I think pre-processing the image is precisely what should be in the multi-modal processor. Users should be able to directly pass images into vLLM instead of having to manually run pre-processing outside of it. Do tell me if this doesn't fit your use case though!

christian-pinto commented:
> Hmm, I think pre-processing the image is precisely what should be in the multi-modal processor. Users should be able to directly pass images into vLLM instead of having to manually run pre-processing outside of it. Do tell me if this doesn't fit your use case though!

As far as I understand, right now the multi-modal processor is invoked in vLLM before inference and its output is fed to the model. One model inference is executed per pre-processed image(s). In our use case the model is not processing the entire image at once, but in patches of fixed size. IN the pre-processing phase we split the image (and apply other transformations), and generate a number of patches depending on the size of the image itself. We then run one model inference for each patch and then post-process the output from each inference to re-build the output image. At the moment I don't think I can replicate our use-case with vLLM. Correct me if I am wrong of course.

christian-pinto committed:
Added PrithviGeoSpatialMAE to model registry tests

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented:
> As far as I understand, right now the multi-modal processor is invoked in vLLM before inference and its output is fed to the model. One model inference is executed per pre-processed image(s). In our use case the model is not processing the entire image at once, but in patches of fixed size. IN the pre-processing phase we split the image (and apply other transformations), and generate a number of patches depending on the size of the image itself. We then run one model inference for each patch and then post-process the output from each inference to re-build the output image. At the moment I don't think I can replicate our use-case with vLLM. Correct me if I am wrong of course.

Thanks for the explanation. Roughly how many patches are there? I'm wondering whether we can still consider one image (corresponding to multiple patches) per prompt, and batch together the patches inside the model.

christian-pinto commented:
> Thanks for the explanation. Roughly how many patches are there? I'm wondering whether we can still consider one image (corresponding to multiple patches) per prompt, and batch together the patches inside the model.

I see what you mean, the model could receive all the patches at once (so far I have seen 1 to 4 per image but I am not sure about an upper limit) from the input processor and perform the batching internally. So this would still be one request for vLLM. It is definitely possible, however we would need to add a post-processing step as well where the output image is reconstructed. In our specific case this requires also some information from the input image because the format we use (geotiff) contains localization data that needs to be also in the output image.

In general, in the road towards proper support for non-language models, most models would benefit from having an output processor that generates data in the correct modality and in a way this is happening already with the detokenization to generate text output. 

This sounds like something bigger that we could address in a follow-up?

christian-pinto committed:
Avoid retrieving tokenizer in _process_multimodal when skip_tokenizer_init=True

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented:
> This sounds like something bigger that we could address in a follow-up?

Sounds good. In the meantime, can you add a separate example script under `examples` directory showing how the run the model in vLLM?

christian-pinto committed:
_process_multimodal: skip getting lora tokenizer if the model is initialized with skip_tokenizer_init=True

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
PrithviGeospatialMAE example script

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
@DarkLight1337 I have done the following changes:

- Added the model to the multimodal registry test with `is_available_online=False`
- As per our discussion I am skipping the other tests as we do not have a HF processor at the moment
- Added a script how to perform a full inference using vLLM. This also shows an example image transformations that would "normally" be part of the HF Processor.

christian-pinto committed:
Fixed formatting issues

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
More formatting issues

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
Hey @DarkLight1337 any further comments?

I have checked the failing tests but they seem to be unrelated to my changes:
- read timeouts from huggingface.co
https://buildkite.com/vllm/fastcheck/builds/13639#0194efb0-9b49-459c-8bb5-a18bbd0b0be7/205-13581

- Failures in some video and audio tests for the openai interface
https://buildkite.com/vllm/fastcheck/builds/13639#0194efb0-9b4e-4782-a409-01681ecc4739/6-11058

Please, let me know any further step or additional change required.

DarkLight1337 reviewed (approved):
Sorry I was busy, it looks good to me now. The failing tests are due to HF timeout and can be ignored.

christian-pinto commented:
Awesome thanks, I have just realised there's a typing hint missing. Let me push it then.

christian-pinto committed:
Fixed PoolingOutput typing

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Fixed PoolingOutput typing style

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Reverting PoolingOutput changes as the current PrithviGeospatial model only returns one Tensor

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
@DarkLight1337 I am done with my changes. Not sure how auto merging works will all the failing tests (timeouts form HF)

christian-pinto committed:
Avoid using CUDA graph for non-text Pooling models

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on vllm/attention/backends/placeholder_attn.py:
```diff
@@ -316,9 +316,14 @@ def build(self, seq_lens: List[int], query_lens: List[int],
+        # Some input builders such as ModelInputForCPUBuilder do not have the
+        # "inter_data_list" attribute.
+        # Let's check inter_data_list exists before we reference it.
+        if hasattr(self.input_builder, "inter_data_lis"):
```
```suggestion
        if hasattr(self.input_builder, "inter_data_list"):
```

Just noticed a typo here

christian-pinto replied:
great catch!

christian-pinto replied:
Fixed with last series of commits

christian-pinto committed:
Fixed typo in placeholder_attn

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Parse model weights

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

Isotr0py commented on examples/offline_inference/prithvi_geospatial_mae.py:
```diff
@@ -0,0 +1,530 @@
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser("MAE run inference", add_help=False)
+
+    parser.add_argument(
+        "--data_file",
+        type=str,
+        default="./India_900498_S2Hand.tif",
+        help="Path to the file.",
+    )
```
Seems that users have to download the example image manually before running the script. Can we download the sentinel-2 image during runtime?

christian-pinto replied:
The images that we use for this example are indeed to be downloaded manually but it is a single tiff file that we host on hugging face. I see it as a minimal disruption for the user at this stage. Next I will work on better automated execution of the model so that we can have a full functional test in vLLM.

Isotr0py reviewed (approved):
Thanks for bringing this model to vLLM!

christian-pinto committed:
Cleanup

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
The last failing test is related to commit ce26b16 and it seems not linked to this PR. Unless I am missing something.

thesteve0 commented:
Awesome that this is V2 of Prithvi as well
